linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Types.scala

```diff
@@ -0,0 +1,174 @@
+    def apply(heapType: HeapType): RefType = RefType(false, heapType)
+
+    /** Builds a non-nullable `(ref typeID)` for the given `typeID`. */
+    def apply(typeID: TypeID): RefType = apply(HeapType(typeID))
+
+    /** Builds a nullable `(ref null heapType)` for the given `heapType`. */
+    def nullable(heapType: HeapType): RefType = RefType(true, heapType)
+
+    /** Builds a nullable `(ref null typeID)` for the given `typeID`. */
+    def nullable(typeID: TypeID): RefType = nullable(HeapType(typeID))
+
+    /** `(ref any)`. */
+    val any: RefType = apply(HeapType.Any)
+
+    /** `(ref null any)`, i.e., `anyref`. */
+    val anyref: RefType = nullable(HeapType.Any)
+
+    /** `(ref func)`. */
+    val func: RefType = apply(HeapType.Func)
+
+    /** `(ref null func)`, i.e., `funcref`. */
+    val funcref: RefType = nullable(HeapType.Func)
+
+    /** `(ref extern)`. */
+    val extern: RefType = apply(HeapType.Extern)
+
+    /** `(ref null extern)`, i.e., `externref`. */
+    val externref: RefType = nullable(HeapType.Extern)
+
+    /** `(ref null exn)`, i.e., `exnref`. */
+    val exnref: RefType = nullable(HeapType.Exn)
+
+    /** `(ref null none)`, i.e., `nullref`. */
+    val nullref: RefType = nullable(HeapType.None)
+  }
+
+  /** A WebAssembly `heaptype`. */
+  sealed abstract class HeapType
+
+  object HeapType {
+
+    /** Reference to a named composite type. */
+    final case class Type(typeID: TypeID) extends HeapType
+
+    /** A WebAssembly `absheaptype`. */
+    sealed abstract class AbsHeapType(val textName: String,
+        val nullableRefTextName: String, val binaryCode: Byte)
+        extends HeapType
+
+    case object Func extends AbsHeapType("func", "funcref", 0x70)
+    case object Extern extends AbsHeapType("extern", "externref", 0x6F)
+    case object Any extends AbsHeapType("any", "anyref", 0x6E)
+    case object Eq extends AbsHeapType("eq", "eqref", 0x6D)
+    case object Array extends AbsHeapType("array", "arrayref", 0x6A)
+    case object Exn extends AbsHeapType("exn", "exnref", 0x69)
+    case object Struct extends AbsHeapType("struct", "structref", 0x6B)
+    case object None extends AbsHeapType("none", "nullref", 0x71)
+    case object NoExtern extends AbsHeapType("noextern", "nullexternref", 0x72)
+    case object NoFunc extends AbsHeapType("nofunc", "nullfuncref", 0x73)
+    case object NoExn extends AbsHeapType("noexn", "nullexnref", 0x74)
+
+    def apply(typeID: TypeID): HeapType.Type =
+      HeapType.Type(typeID)
+  }
+
+  /** A WebAssembly `rectype`. */
+  final case class RecType(subTypes: List[SubType])
+
+  object RecType {
+
+    /** Builds a `rectype` with a single `subtype`. */
+    def apply(singleSubType: SubType): RecType =
+      RecType(singleSubType :: Nil)
+  }
+
+  /** A WebAssembly `subtype` with an associated name. */
+  final case class SubType(
+      id: TypeID,
+      originalName: OriginalName,
+      isFinal: Boolean,
+      superType: Option[TypeID],
```

Comment:
What is this? I cannot find any reference to it in the spec, but the TextWriter / BinaryWriter seem to use it.

Reply:
It's the `typeidx*` of the spec at https://webassembly.github.io/gc/core/syntax/types.html#recursive-types

There is an additional restriction that there is at most one, and it is interpreted as a super type so we gave it a better name here.

Note that the `id` also ultimately corresponds to a `typeidx`, but that one is implicitly allocated by the order of definitions in the types section.